File: cses/08_string_algorithms/2110_substring-distribution/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    s = data[0]                      # bytes; lặp qua cho ra các int
    n = len(s)

    # Các mảng của suffix automaton (SAM): len, suffix link, cạnh chuyển (dict)
    max_states = 2 * n + 5
    sa_len = [0] * max_states
    sa_link = [-1] * max_states
    sa_next = [None] * max_states
    sa_next[0] = {}                  # state khởi tạo 0
    size = 1
    last = 0

    # Xây dựng SAM online: thêm lần lượt từng ký tự của s
    for ch in s:                     # ch là int (giá trị byte)
        cur = size
        size += 1
        sa_len[cur] = sa_len[last] + 1
        sa_next[cur] = {}
        # Đi theo suffix link, thêm cạnh chuyển tới state mới
        p = last
        while p != -1 and ch not in sa_next[p]:
            sa_next[p][ch] = cur
            p = sa_link[p]
        if p == -1:
            # Không có cạnh trùng: suffix link trỏ về state khởi tạo
            sa_link[cur] = 0
        else:
            q = sa_next[p][ch]
            if sa_len[p] + 1 == sa_len[q]:
                # Cạnh liên tục: dùng thẳng q làm suffix link
                sa_link[cur] = q
            else:
                # Tách state (clone) để giữ tính tối thiểu của automaton
                clone = size
                size += 1
                sa_len[clone] = sa_len[p] + 1
                sa_next[clone] = dict(sa_next[q])
                sa_link[clone] = sa_link[q]
                # Chuyển các cạnh trỏ tới q sang clone
                while p != -1 and sa_next[p].get(ch) == q:
                    sa_next[p][ch] = clone
                    p = sa_link[p]
                sa_link[q] = clone
                sa_link[cur] = clone
        last = cur

    # Mảng hiệu: mỗi state cộng 1 vào các độ dài [len[link]+1, len]
    diff = [0] * (n + 2)
    for v in range(1, size):
        L = sa_len[sa_link[v]] + 1
        R = sa_len[v]
        diff[L] += 1
        diff[R + 1] -= 1

    # Prefix sum: giá trị tại L là số xâu con phân biệt có độ dài L
    out = []
    run = 0
    for L in range(1, n + 1):
        run += diff[L]
        out.append(run)

    sys.stdout.write(' '.join(map(str, out)))
    sys.stdout.write('\n')


main()

File: cses/08_string_algorithms/2110_substring-distribution/solution.cpp
#include <array>
#include <iostream>
#include <string>
#include <vector>

// Một state của suffix automaton (SAM):
// - length: độ dài xâu con dài nhất mà state biểu diễn
// - link: suffix link
// - next: cạnh chuyển theo từng ký tự a-z
struct State {
    int length = 0;
    int link = -1;
    std::array<int, 26> next{};

    State() {
        next.fill(-1);
    }
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string text;
    std::cin >> text;

    // Xây dựng SAM online: bắt đầu với state khởi tạo 0
    std::vector<State> automaton;
    automaton.reserve(2 * text.size());
    automaton.emplace_back();
    int last = 0;

    for (const char character : text) {
        const int letter = character - 'a';

        // Tạo state mới ứng với tiền tố dài thêm 1 ký tự (state dài nhất mới)
        const int current = static_cast<int>(automaton.size());
        automaton.emplace_back();
        automaton[current].length = automaton[last].length + 1;

        // Đi theo suffix link, thêm cạnh chuyển tới state mới
        int state = last;
        while (state != -1 && automaton[state].next[letter] == -1) {
            automaton[state].next[letter] = current;
            state = automaton[state].link;
        }
        if (state == -1) {
            // Không có cạnh trùng: suffix link trỏ về state khởi tạo
            automaton[current].link = 0;
        } else {
            const int child = automaton[state].next[letter];
            if (automaton[state].length + 1 == automaton[child].length) {
                // Cạnh liên tục: dùng thẳng child làm suffix link
                automaton[current].link = child;
            } else {
                // Tách state (clone) để giữ tính tối thiểu của automaton
                const int clone = static_cast<int>(automaton.size());
                automaton.push_back(automaton[child]);
                automaton[clone].length = automaton[state].length + 1;
                // Chuyển các cạnh trỏ tới child sang clone
                while (state != -1 &&
                       automaton[state].next[letter] == child) {
                    automaton[state].next[letter] = clone;
                    state = automaton[state].link;
                }
                automaton[child].link = clone;
                automaton[current].link = clone;
            }
        }
        last = current;
    }

    // Mảng hiệu: mỗi state v cộng 1 cho khoảng độ dài [len[link]+1, len[v]]
    std::vector<int> difference(text.size() + 2, 0);
    for (std::size_t state = 1; state < automaton.size(); ++state) {
        const int left =
            automaton[static_cast<std::size_t>(automaton[state].link)].length + 1;
        const int right = automaton[state].length;
        ++difference[static_cast<std::size_t>(left)];
        --difference[static_cast<std::size_t>(right + 1)];
    }

    // Prefix sum: giá trị tại L là số xâu con phân biệt có độ dài L
    int count = 0;
    for (std::size_t length = 1; length <= text.size(); ++length) {
        count += difference[length];
        if (length > 1) {
            std::cout << ' ';
        }
        std::cout << count;
    }
    std::cout << '\n';
    return 0;
}
